src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AccountControllerTests.cs

```diff
@@ -0,0 +1,743 @@
+        }
+
+        [Fact]
+        public void Login_UserIsNull_ReturnsViewWithError()
+        {
+            // Arrange
+            Uri returnUrl = null;
+            LoginViewModel model = GetLoginViewModel();
+            this._accountServiceMock.Setup(x => x.FindByEmailAsync(model.Email)).ReturnsAsync((ApplicationUser)null);
+            using var controller = new AccountController(this._accountServiceMock.Object, this._loggerMock.Object);
+
+            // Act
+            var result = controller.Login(model, returnUrl);
+
+            // Assert
+            var allErrors = controller.ModelState.Values.SelectMany(v => v.Errors);
+            Assert.Single(allErrors);
+            Assert.Contains(allErrors, x => x.ErrorMessage == @EmailNotFound);
+            var viewResult = Assert.IsType<Task<IActionResult>>(result);
+        }
+
+        [Fact]
+        public void Login_EmailNotConfirmed_ReturnsViewWithError()
+        {
+            // Arrange
+            Uri returnUrl = null;
+            LoginViewModel model = GetLoginViewModel();
+            this._accountServiceMock.Setup(x => x.FindByEmailAsync(model.Email))
+                                    .ReturnsAsync(new ApplicationUser() { EmailConfirmed = false });
+            using var controller = new AccountController(this._accountServiceMock.Object, this._loggerMock.Object);
+
+            // Act
+            var result = controller.Login(model, returnUrl);
+
+            // Assert
+            var allErrors = controller.ModelState.Values.SelectMany(v => v.Errors);
+            Assert.Single(allErrors);
+            Assert.Contains(allErrors, x => x.ErrorMessage == @EmailMustBeConfirmed);
+            var viewResult = Assert.IsType<ViewResult>(result.Result);
+            Assert.IsType<LoginViewModel>(viewResult.ViewData.Model);
+        }
+
+        [Fact]
+        public void Login_ResultIsSucceeded_RedirectToLocal()
+        {
+            // Arrange
+            Uri returnUrl = null;
+            LoginViewModel model = GetLoginViewModel();
+            this._accountServiceMock.Setup(x => x.FindByEmailAsync(model.Email))
+                                    .ReturnsAsync(new ApplicationUser() { EmailConfirmed = true });
+            this._accountServiceMock.Setup(x => x.PasswordSignInAsync(It.IsAny<string>(), model.Password, model.RememberMe))
+                                   .ReturnsAsync(SignInResult.Success);
+            using var controller = new AccountController(this._accountServiceMock.Object, this._loggerMock.Object);
+            var mockUrlHelper = new Mock<IUrlHelper>();
+            mockUrlHelper.Setup(x => x.Action(It.IsAny<UrlActionContext>())).Returns("callbackUrl");
+            controller.Url = mockUrlHelper.Object;
+
+            // Act
+            var result = controller.Login(model, returnUrl);
+
+            // Assert
+            var viewResult = Assert.IsType<RedirectToActionResult>(result.Result);
```

Comment:
As above, we should consider awaiting that action and not waiting synchronously for its completion.